go/libraries/doltcore/sqle/rowexec/lookup_join.go

```diff
@@ -0,0 +1,327 @@
+		split = 1
+	} else {
+		split = sourceSch.GetPKCols().Size()
+	}
+
+	// schMappings tell us where to look for key fields. A field will either
+	// be in the source key tuple (< split), source value tuple (>=split),
+	// or in the literal tuple (-1).
+	srcMapping := make(val.OrdinalMapping, len(keyExprs))
+	var litMappings val.OrdinalMapping
+	var litTypes []val.Type
+	for i, e := range keyExprs {
+		switch e := e.(type) {
+		case *expression.GetField:
+			// map the schema order index to the physical storage index
+			col := sourceSch.GetAllCols().NameToCol[e.Name()]
+			if col.IsPartOfPK {
+				srcMapping[i] = sourceSch.GetPKCols().TagToIdx[col.Tag]
+			} else if keyless {
+				// Skip cardinality column
+				srcMapping[i] = split + 1 + sourceSch.GetNonPKCols().TagToIdx[col.Tag]
+			} else {
+				srcMapping[i] = split + sourceSch.GetNonPKCols().TagToIdx[col.Tag]
+			}
+		case *expression.Literal:
+			srcMapping[i] = -1
+			litMappings = append(litMappings, i)
+			litTypes = append(litTypes, tgtKeyDesc.Types[i])
+		}
+	}
+	litDesc := val.NewTupleDescriptor(litTypes...)
+	litTb := val.NewTupleBuilder(litDesc)
+	for i, j := range litMappings {
+		tree.PutField(ctx, src.NodeStore(), litTb, i, keyExprs[j].(*expression.Literal).Value())
+	}
+
+	var litTuple val.Tuple
+	if litDesc.Count() > 0 {
+		litTuple = litTb.Build(src.Pool())
+	}
+
+	return &lookupMapping{
+		split:      split,
+		srcMapping: srcMapping,
+		litTuple:   litTuple,
+		litKd:      litDesc,
+		srcKd:      src.KeyDesc(),
+		srcVd:      src.ValDesc(),
+		targetKb:   val.NewTupleBuilder(tgtKeyDesc),
+		ns:         src.NodeStore(),
+		pool:       src.Pool(),
+	}
+}
+
+// valid returns whether the source and destination key types
+// are type compatible
+func (m *lookupMapping) valid() bool {
+	var litIdx int
+	for to := range m.srcMapping {
+		from := m.srcMapping.MapOrdinal(to)
+		var desc val.TupleDesc
+		if from == -1 {
+			desc = m.litKd
+			// literal offsets increment sequentially
+			from = litIdx
+			litIdx++
+		} else if from < m.split {
+			desc = m.srcKd
+		} else {
+			// value tuple, adjust offset
+			desc = m.srcVd
+			from = from - m.split
+		}
+		if desc.Types[from].Enc != m.targetKb.Desc.Types[to].Enc {
+			return false
+		}
+	}
+	return true
+}
+
+func (m *lookupMapping) dstKeyTuple(ctx context.Context, srcKey, srcVal val.Tuple) (val.Tuple, error) {
+	var litIdx int
+	for to := range m.srcMapping {
+		from := m.srcMapping.MapOrdinal(to)
+		var tup val.Tuple
+		var desc val.TupleDesc
+		if from == -1 {
+			tup = m.litTuple
+			desc = m.litKd
+			// literal offsets increment sequentially
+			from = litIdx
+			litIdx++
+		} else if from < m.split {
+			desc = m.srcKd
+			tup = srcKey
+		} else {
+			// value tuple, adjust offset
+			tup = srcVal
+			desc = m.srcVd
+			from = from - m.split
+		}
+
+		if desc.Types[from].Enc == m.targetKb.Desc.Types[to].Enc {
+			m.targetKb.PutRaw(to, desc.GetField(from, tup))
+		} else {
+			// TODO how to do GMS consistent conversion?
```

Comment:
Resolve this TODO

Reply:
I also think this is out of scope. The observation here is that we're just avoiding the optimization when we do a lookup from ex: string->int. These are rare and subject to a big list of subtle type conversion specificities.

Reply:
I'll expand the TODO though, to be more specific

Reply:
made this an error also, we reject this path upstream